app/lib/smooch_nlu.rb

```diff
@@ -55,21 +58,33 @@ def self.menu_option_from_message(message, options)
         }
       }
       response = Bot::Alegre.request_api('get', '/text/similarity/', params)
-      best_result = response['result'].to_a.sort_by{ |result| result['_score'] }.last
-      unless best_result.nil?
-        option = options.find{ |o| !o['smooch_menu_option_id'].blank? && o['smooch_menu_option_id'] == best_result.dig('_source', 'context', 'menu_option_id') }
+      # Sort the results from best to worst
+      # sorted_options = response['result'].to_a.sort_by{ |result| result['_score'] }.reverse()
+
+      # Get the menu_option_id of all results returned
+      option_counts = response['result'].to_a.map{|o| o.dig('_source', 'context', 'menu_option_id')}
+      # Count how many of each menu_option_id we have and sort (high to low)
+      option_counts = option_counts.group_by(&:itself).transform_values(&:count).sort_by{|_k,v| v}.reverse()
+
+      # Take option menu that is most prevalent in the results and exists in `options`
+      option_counts.each do | r, _ |
+        option = options.find{ |o| !o['smooch_menu_option_id'].blank? && o['smooch_menu_option_id'] == r }
+        break if !option.nil?
       end
-      log = {
-        version: "0.1", # Update if schema changes
-        datetime: DateTime.current,
-        team_slug: team_slug,
-        user_query: message,
-        alegre_query: params,
-        alegre_response: response,
-        selected_option: option
-      }
-      Rails.logger.info("[Smooch NLU] [Menu Option From Message] #{log.to_json}")
+
+      # FIXME: Deal with ties (i.e., where two options have an equal count)
     end
+    # In all cases log for analysis
```

Comment:
I'd like @caiosba input here. Ideally, I'd like to do this logging for any message where we expected a menu option and didn't get one. That will allow us to analyze the intents we currently fail to respond to. I don't know, however, how 'heavy' an operation this is or if the `menu_option_from_message(message, options)` is a heavy traffic function such that this would have an impact on performance?

Reply:
It's OK to log for all iterations here.

Reply:
Great. Thanks!